src/TPPCore.ChatProviders/Providers/Twitch/TwitchIrcProvider.cs

```diff
@@ -23,25 +26,105 @@ public TwitchIrcProvider()
+            ircClient.CommandHandlers.AddOrCombine("USERNOTICE", UserNoticeEventHandler);
+            ircClient.CommandHandlers.AddOrCombine("WHISPER", WhisperEventHandler);
         }
 
         override protected ChatUser getMessageSender(Message message)
         {
-            return new ChatUser() {
+            return new ChatUser()
+            {
                 UserId = message.Tags.ContainsKey("user-id")
                     ? message.Tags["user-id"] : null,
                 Username = message.Tags.ContainsKey("login")
                     ? message.Tags["login"]
-                    : (message.Prefix.ClientId != null
-                        ? message.Prefix.ClientId.Nickname : null),
+                    : (message.Prefix.ClientId?.Nickname),
                 Nickname = message.Tags.ContainsKey("display-name")
                     ? message.Tags["display-name"] : null
             };
         }
 
-        private Task userNoticeEventHandler(IrcClient client, Message message)
+        private Task MessageEventHandler(IrcClient client, Message message)
+        {
+            var chatMessage = new ChatMessage()
+            {
+                ClientName = ClientName,
+                ProviderName = ProviderName,
+                TextContent = message.TrailingParameter,
+                Channel = message.TargetLower,
+                Sender = getMessageSender(message),
+                IsSelf = isMessageSelf(message)
+            };
+            message.Tags.ToList().ForEach(item => chatMessage.Meta.Add(item));
+            message.Tags.TryGetValue("emotes", out string value);
+
+            if (value != "" && value != null)
+            {
+                List<EmoteOccurance> emoteOccurances = new List<EmoteOccurance> { };
+                Dictionary<string, Tuple<int, string>> keyValuePairs = new Dictionary<string, Tuple<int, string>> { };
+                List<string> emotes = value.Split('/').ToList(); // / seperates emote types
+                foreach (string emote in emotes)
+                {
+                    int.TryParse(emote.Split(':')[0], out int emoteId); // : seperates the ID from the indexes
+                    string indexes = emote.Split(':')[1];
+                    if (emote.Contains(',')) // , seperates emote occurances
+                    {
+                        List<string> occurances = indexes.Split(',').ToList();
+                        foreach (string occurance in occurances)
+                        {
+                            int.TryParse(occurance.Split('-')[0], out int firstIndex); // - seperates the start index from the end index
+                            int.TryParse(occurance.Split('-')[1], out int secondIndex);
+                            string name = string.Empty;
+                            for (int i = firstIndex; i <= secondIndex; i++)
+                            {
+                                name += chatMessage.TextContent[i];
+                            }
```

Comment:
I think you want to use `string.Substring(index, length)`?

Checking the result of `int.TryParse` and skipping this occurrence if it fails might be a good idea too but not necessary.